source/mir/random/nonuniform.d

```diff
@@ -0,0 +1,246 @@
+/**
+Non-uniform distributions.
+
+License: $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Sebastian Wilzbach
+*/
+
+module mir.random.nonuniform;
+
+import std.traits : isNumeric;
+import std.stdio;
+
+/**
+Setup a _normal distribution sampler.
+
+Params:
+    T = floating-point type that should be sampled
+    R = unsigned UIntType of the random generator
+
+Returns:
+    A $(LREF Ziggurat) sampler that can be called to sample from the distribution.
+*/
+auto normal(T, R = uint)()
+{
+    import mir.internal.math : exp, log, sqrt;
+
+    auto pdf    = (T x) => cast(T) exp(T(-0.5) * x * x);
+    auto invPdf = (T x) => cast(T) sqrt(T(-2) * log(x));
+
+    // values from [Marsaglia00]
+    T rightEnd = 3.442619855899;
+
+    /// generate x for the last block (aka tail-block)
+    enum fallback = q{
+        T fallback(RNG)(bool isPositive, ref RNG gen) const
+        {
+            import std.random : uniform;
+            import std.math : log;
+            T x, y, u;
+            do
+            {
+                u = uniform!("[]", T, T)(0, 1, gen);
+                x = -log(u) / fs[$ - 1];
+                u = uniform!("[]", T, T)(0, 1, gen);
+                y = -log(u);
+            }
+            while(y + y < x * x);
+            return isPositive ? rightEnd + x : -rightEnd - x;
+        }
+    };
+
+    return Ziggurat!(T, fallback, R, true)(pdf, invPdf, 128, rightEnd, T(9.91256303526217e-3));
+}
+
+/**
+Setup an _exponential distribution sampler.
+
+Params:
+    T = floating-point type that should be sampled
+    R = unsigned UIntType of the random generator
+
+Returns:
+    A $(LREF Ziggurat) sampler that can be called to sample from the distribution.
+*/
+auto exponential(T, R = uint)()
+{
+    import mir.internal.math : exp, log;
+
+    auto pdf    = (T x) => cast(T) exp(-x);
+    auto invPdf = (T x) => cast(T) -log(x);
+
+    // values from [Marsaglia00]
+    enum fallback = q{
+        T fallback(RNG)(ref RNG gen) const
+        {
+            import std.random : uniform;
+            auto u = uniform!("[]", T, T)(0, 1, gen);
+            return 7.69711 - u;
+        }
+    };
+
+    return Ziggurat!(T, fallback, R, false)(pdf, invPdf, 256, T(7.697117470131487), T(3.949659822581572e-3));
+}
+
+/**
+Ziggurat algorithm for generating a random variable from a montone, decreasing density.
+
+Complexity: O_avg: O(1)
+
+References:
+    Marsaglia, George, and Wai Wan Tsang. "The ziggurat method for generating random variables."
+    Journal of statistical software 5.8 (2000): 1-7.
+*/
+struct Ziggurat(T, string _fallback, R = uint, bool bothSides)
+    if (isNumeric!T)
+{
+
+private:
+
+    /// monotone decreasing probability density function
+    /// constants don't matter
+    T function(T x) pdf;
+    T function(T x) invPdf;
+
+    /// precalculate difference x_i / x_{i+1}
+    T[] xDiv;
+    /// precalculate scaling to R.max for x_i
+    T[] xScaled;
+    /// precalculate pdf value for x_i
+    T[] fs;
```

Comment:
"An Improved Ziggurat Method to Generate Normal Random Samples" postulates that saving the `fs` isn't necessary to which I couldn't follow yet.

> DRanU() returns a uniform random number, U (0, 1), and IRanU() returns 32-bit
unsigned random integer. DRanNormalTail is implemented as a separate function:
it gets called only rarely, so that efficiency does not matter. For the same reason, it
is not necessary to avoid a call to exp() when checking for the wedges (this could
be achieved by precomputing the function values f (xi). (page 7)